Wake delivery cannot degrade a dead route, so it retries forever

## Context

Observed live in the containerized Memory Core on 2026-08-01, triggered by an ordinary `add_message` broadcast. From `/app/.neo-ai-data/logs/mc-server-2026-08-01.log` inside `neo-local-agent-os-mc-server-1`:

```
00:07:25 [ERROR] WebhookDeliveryService: Network error delivering to WAKE_SUB:36342c42-…: fetch failed. Attempt 1/4
00:07:26 [ERROR] … Attempt 2/4
00:07:28 [ERROR] … Attempt 3/4
00:07:35 [ERROR] … Attempt 4/4
00:07:35 [WARN]  WebhookDeliveryService: 3 consecutive failures for WAKE_SUB:36342c42-…. Marking degraded.
00:07:35 [WARN]  WebhookDeliveryService: Cannot degrade WAKE_SUB:36342c42-…, node not found in Graph.
```

The delivery failure itself is expected and not the subject of this ticket — no wake receiver is currently listening, which is `#16233`'s scope. The defect is the last line: the service decided to degrade the route and **could not**, so nothing changed and the next message repeats the whole sequence.

Observation separated from inference: the log lines above are observed. The mechanism in The Problem is derived from source plus a within-run control, and its necessity is not yet directly measured — the falsifier is named in the ACs.

## The Problem

**The node is not missing. The read path cannot see it.**

The within-run control is decisive: in that same flush, the *same process* successfully read the same subscription — it had to, in order to resolve the `url` and `signingKey` it delivered against. One read succeeded and one returned nothing, milliseconds apart, on the same node. So "node not found" is not about the node's existence.

The two reads use different APIs:

- `WakeSubscriptionService` reads through `GraphService.db.nodes.get(id)` directly, with an explicit comment at `ai/services/memory-core/WakeSubscriptionService.mjs:854` — *"Access `GraphService.db.nodes.get` directly because `GraphService.getNode` filters out custom properties"* — and again at `:1373`.
- `WebhookDeliveryService._markDegraded` calls `GraphService.getNode({id: subscriptionId})` at `ai/services/memory-core/WebhookDeliveryService.mjs:158`.

`GraphService.getNode` applies a row-level-security re-check at its return boundary (`GraphService.mjs:845-867`). `isRlsVisible` (`:27-45`) returns true only when the owner key is null, the normalized owner matches the requester, the entity is shared, or visibility is `team`. `resolveRlsUserId` (`:62-65`) resolves the requester from the request-bound `RequestContextService`.

The coalescing flush is a **background timer**, not an MCP request. With no bound request context there is no requester id, the subscription is owner-stamped and not shared, and every branch of `isRlsVisible` evaluates false — so `getNode` returns `null` and the degrade is skipped. `WakeSubscriptionService` already knows this and routes around it; `WebhookDeliveryService` does not.

**Falsified competing hypothesis.** Stored `userId` is inconsistent across subscription vintages — some records carry `@neo-opus-ada`, others `neo-opus-ada`. That looked like the cause and is not: `isRlsVisible` normalizes the owner key before comparing (`GraphService.mjs:41`), explicitly to match an agent's own nodes regardless of stored form. Recording it so the next reader does not re-derive and stop there.

**Consequence.** A permanently dead route is never marked degraded, so `consecutiveFailures` keeps climbing past 3 and re-fires `_markDegraded` on every subsequent message, each time preceded by 4 delivery attempts with backoff (`:145-153`). Every A2A message to that seat costs four failed network round trips forever. The only signal is a WARN in a file nobody tails — `docker logs` does not carry it.

## The Architectural Reality

| Surface | Role |
|---|---|
| `ai/services/memory-core/WebhookDeliveryService.mjs:155-176` | `_markDegraded`; the RLS-scoped read that fails |
| `ai/services/memory-core/WebhookDeliveryService.mjs:145-153` | `_recordConsecutiveFailure`; re-fires the degrade on every message past the threshold |
| `ai/services/memory-core/GraphService.mjs:845-867` | `getNode`; RLS re-check at the return boundary |
| `ai/services/memory-core/GraphService.mjs:27-45` | `isRlsVisible`; all four branches false for an owner-stamped node with no requester |
| `ai/services/memory-core/WakeSubscriptionService.mjs:854`, `:1373` | the sibling that already bypasses `getNode`, with the reason in a comment |
| `ai/services/memory-core/CoalescingEngineService.mjs:518` | `Unknown harnessTarget '<t>'; dropping digest` — where a `degraded` value would land |

This is a **service-boundary seam**, not a logic bug in either component. `getNode`'s RLS is behaving correctly and must not be weakened; the caller is simply on the wrong side of a request boundary. The fix belongs to the caller.

## The Fix

1. `_markDegraded` reads the subscription through the same context-free path its sibling uses, rather than the request-scoped `getNode`. `GraphService.getNodeRecord`-style properties-returning reads exist (`GraphService.mjs:869-879`) but carry the identical RLS re-check, so they are not a fix — the background-writer path is what needs naming.
2. Give the background writer an explicit, named accessor rather than each service reaching into `GraphService.db.nodes` ad hoc. Two call sites already do it by hand with an explanatory comment; a third arriving by copy-paste is how this recurs.
3. `_recordConsecutiveFailure` must not re-enter the 4-attempt delivery cycle for a route already known-dead. Today the threshold triggers a degrade that silently fails, and nothing caps the repetition.
4. Decide and document what `harnessTarget: 'degraded'` **is**. It is written at `:162` but appears in no enum: the `manage_wake_subscription` tool schema accepts only `mcp-notifications | a2a-webhook | bridge-daemon | disabled | none`. It currently "works" only by falling through `CoalescingEngineService`'s unknown-target branch, which drops the digest with a WARN. Working by accident through an unknown-value fallback is not a state model — either add it to the enum as a first-class state, or record degradation outside `harnessTarget` and leave the routing field to routing.

## Contract Ledger

| Target Surface | Source of Authority | Proposed Behavior | Fallback | Docs | Evidence |
|---|---|---|---|---|---|
| `WebhookDeliveryService._markDegraded` node read | `WakeSubscriptionService.mjs:854` precedent | context-free read that succeeds from a background flush | none; a failed degrade must log at ERROR, not WARN | this ticket | a forced-failure route reaches a degraded state within the threshold |
| Background-writer graph accessor | `GraphService` RLS contract | one named accessor for context-free writers | direct `db.nodes.get` remains, but only behind that accessor | JSDoc on the accessor | no service reaches into `GraphService.db.nodes` directly |
| `harnessTarget` value space | `manage_wake_subscription` tool enum | `degraded` is either in the enum or not written to this field | out-of-enum values rejected at write | tool schema + `WakeSubscriptionService` JSDoc | `list` never returns a `harnessTarget` outside the documented enum |
| Retry behaviour for a known-dead route | `_recordConsecutiveFailure` | bounded; no unbounded repetition of the 4-attempt cycle | — | this ticket | attempt count for a permanently dead route is bounded across N messages |

## Decision Record impact

`none`. This restores intended behaviour at a service seam; it does not alter the RLS model, which is working as designed.

## Acceptance Criteria

- [ ] A subscription whose endpoint refuses connections reaches a degraded state, verified by reading the subscription back after the failure threshold — not by reading the log line.
- [ ] The necessity probe is run and recorded: log the resolved RLS requester at flush time and confirm it is absent. If a requester *is* bound, the mechanism in The Problem is wrong and this ticket is re-derived before any fix lands.
- [ ] A test drives the degrade path with **no bound request context** and fails on current `dev`. A test that passes a context would heal the defect in setup and prove nothing.
- [ ] Delivery attempts against a permanently dead route are bounded across repeated messages; asserted by counting attempts over N sends, not by inspecting the threshold constant.
- [ ] `harnessTarget` never holds a value outside the documented tool enum, or the enum documents `degraded` as a first-class state. Whichever is chosen is written down once, where both the tool schema and the service can cite it.
- [ ] A failed degrade logs at ERROR. A route silently failing to change state is the condition that made this invisible for as long as it was.
- [ ] No new direct `GraphService.db.nodes` reach-in outside the named accessor.

## Out of Scope

- Provisioning a wake receiver, or the route manifest generator — `#16233`.
- The container-to-host delivery path itself, which is verified working: four delivery attempts left the container against a host URL in the log above.
- Any change to `isRlsVisible`, `resolveRlsUserId`, or the RLS model. The security check is correct; the caller is on the wrong side of it.
- The stored `userId` `@`-prefix inconsistency. Real, but explicitly normalized for and not the cause here.

## Avoided Traps

- **Weakening RLS to make the read succeed.** The obvious one-line "fix", and it would open a cross-tenant read path to close a logging bug.
- **Asserting the userId format mismatch as the cause.** It is the more visible anomaly and it is a red herring; the normalization at `GraphService.mjs:41` rules it out.
- **Testing with a bound request context.** Setup would heal the defect and the test would pass on unfixed code.
- **Trusting `docker logs`.** The mc-server container's stdout carries ~28 boot lines; the real log is a file inside the container. A grep against `docker logs` returns a zero that reads as proof of absence and is pure instrument failure — count lines in the window before trusting any negative here.

## Related

- `#16233` — the missing wake receiver; why this route is dead in the first place
- `#16167` — the migration this surfaced under
- `#16223`, `#16224` — **the same defect class in different subsystems**: retry or suppression state that never advances, so work repeats forever with no state change and no surfaced signal. Three independent instances is a pattern; whoever picks up the third may be looking at a missing shared primitive rather than three local bugs.

Live latest-open sweep: checked the latest 20 open issues at 2026-08-01T00:18:45Z, plus a 30-message A2A in-flight claim scan over the same window; no equivalent ticket and no overlapping `[lane-claim]` found.

Origin Session ID: 56105163-6e66-44b6-8c6f-9e81bc1be08c

Retrieval Hint: `WebhookDeliveryService markDegraded node not found in Graph RLS getNode background flush no request context wake subscription degraded harnessTarget enum`

--- a/ai/services/memory-core/GraphService.mjs
+++ b/ai/services/memory-core/GraphService.mjs
@@ -901,6 +901,57 @@ class GraphService extends Base {
     }
 
     /**
+     * @summary The ONE sanctioned context-free node read, for background writers that run outside any request.
+     *
+     * Every RLS-scoped read ({@link GraphService#getNode}, {@link GraphService#getNodeRecord}) resolves its
+     * requester from the request-bound `RequestContextService`. A background timer — a coalescing flush, a
+     * delivery retry, a maintenance sweep — has no bound request, so `resolveRlsUserId` yields `null`, every
+     * branch of the visibility predicate evaluates false, and the read returns `null` for a node that plainly
+     * exists. The failure is silent and reads as absence: the caller concludes "no such node" and skips the
+     * work it was scheduled to do — a dead wake route that can never be marked degraded re-runs its failing
+     * delivery attempts on every subsequent message, forever.
+     *
+     * Consumers previously reached into `db.nodes.get` by hand to route around this, each with its own
+     * explanatory comment. This is that bypass, named once, so it is greppable and reviewable instead of
+     * arriving by copy-paste.
+     *
+     * **This is NOT an RLS bypass for request-scoped reads.** Calling it from a request path re-opens exactly
+     * the cross-requester cache leak the return-boundary re-check exists to prevent. That misuse is not
+     * mechanically preventable here, so it is made *visible*: a bound request context means the caller is on a
+     * request path and should be using `getNodeRecord`, and the accessor logs a warning naming the writer.
+     * @param {Object} data
+     * @param {String} data.id     Node id.
+     * @param {String} data.writer Name of the background writer, for the misuse warning and audit trails.
+     * @returns {Object|null} `{id, type, properties}`, or `null` only when the node genuinely does not exist.
+     */
+    getUnscopedNodeRecord({id, writer}) {
+        if (!writer) {
+            throw new Error('[GraphService] getUnscopedNodeRecord requires a `writer` name identifying the background writer.');
+        }
+
+        // Guarantee lazy-loading from SQLite triggers if not cached (mirrors the RLS-scoped reads).
+        this.db.getAdjacentNodes(id, 'both');
+
+        const node = this.db.nodes.get(id);
+        if (!node) {
+            return null;
+        }
+
+        // Misuse tripwire: a resolvable requester means a request IS bound, so this caller is on a request
+        // path and must use the RLS-scoped `getNodeRecord` instead. Warn rather than throw — a background
+        // task can legitimately be kicked off during a request — but never let the misuse be silent.
+        if (resolveRlsUserId(this.db.storage?.RequestContextService) !== null) {
+            logger.warn(`[GraphService] getUnscopedNodeRecord called by '${writer}' inside a bound request context; use getNodeRecord on request paths.`);
+        }
+
+        return {
+            id        : node.isRecord ? node.get('id') : node.id,
+            type      : node.isRecord ? node.get('label') : node.label,
+            properties: (node.isRecord ? node.get('properties') : node.properties) || {}
+        };
+    }
+
+    /**
      * @summary Lists full node records of one graph type through the same RLS boundary as single-node reads.
      *
      * This is the sanctioned enumeration companion to {@link GraphService#getNodeRecord}. It keeps

--- a/ai/services/memory-core/WebhookDeliveryService.mjs
+++ b/ai/services/memory-core/WebhookDeliveryService.mjs
@@ -36,6 +36,17 @@ class WebhookDeliveryService extends Base {
     consecutiveFailures = new Map();
 
     /**
+     * Subscription ids degraded during this process lifetime. This is what BOUNDS the attempt count: a
+     * degraded route is never attempted again, so a permanently dead endpoint costs the threshold's attempts
+     * once, not a full retry cycle on every message forever. The persisted
+     * `properties.status` is the durable half and survives restarts; this covers the window before the
+     * caller re-reads the node. {@link WebhookDeliveryService#clearDegraded} is the named resumption path.
+     * @member {Set} degradedSubscriptions
+     * @protected
+     */
+    degradedSubscriptions = new Set();
+
+    /**
      * Per-request timeout injected by the Memory Core entrypoint from
      * `AiConfig.orchestrator.wakeDispatch.attemptTimeoutSeconds`.
      * @member {Number|null}
@@ -66,6 +77,16 @@ class WebhookDeliveryService extends Base {
         const properties = subscription.properties || {};
         const url        = properties.harnessTargetMetadata?.url;
         const signingKey = properties.harnessTargetMetadata?.signingKey;
+
+        // A degraded route is dead until explicitly repaired, so it is not attempted at all. Without this the
+        // failure threshold only decides when to START logging: `_recordConsecutiveFailure` kept counting past
+        // it and re-fired the degrade on every subsequent message, each preceded by a full 4-attempt backoff
+        // cycle. The persisted status is authoritative across restarts; the in-memory set covers
+        // the window before the caller re-reads the node.
+        if (properties.status === 'degraded' || this.degradedSubscriptions.has(subscription.id)) {
+            logger.warn(`WebhookDeliveryService: Subscription ${subscription.id} is degraded; skipping delivery without an attempt.`);
+            return 'skipped';
+        }
 
         if (!url) {
             logger.error(`WebhookDeliveryService: Subscription ${subscription.id} is missing URL.`);
@@ -154,25 +175,54 @@ class WebhookDeliveryService extends Base {
 
     async _markDegraded(subscriptionId) {
         try {
-            // Updating the subscription's harnessTarget to 'degraded'
-            const subscriptionNode = GraphService.getNode({id: subscriptionId});
-            if (subscriptionNode) {
-                const properties = {
-                    ...(subscriptionNode.properties || {}),
-                    harnessTarget: 'degraded'
-                };
+            // Context-free read. This runs from a background delivery flush with no bound request, where an
+            // RLS-scoped read (`GraphService.getNode`) resolves a null requester, fails every visibility
+            // branch, and returns null for a node that plainly exists — the degrade was then skipped and the
+            // route stayed live forever.
+            const record = GraphService.getUnscopedNodeRecord({
+                id    : subscriptionId,
+                writer: 'WebhookDeliveryService._markDegraded'
+            });
 
-                GraphService.upsertNode({
-                    ...subscriptionNode,
-                    properties
-                });
-                logger.info(`WebhookDeliveryService: Subscription ${subscriptionId} marked as degraded.`);
-            } else {
-                logger.warn(`WebhookDeliveryService: Cannot degrade ${subscriptionId}, node not found in Graph.`);
+            if (!record) {
+                // ERROR, not WARN: a route that silently fails to change state is the condition that kept
+                // this invisible. The only signal was a WARN in a file nobody tails.
+                logger.error(`WebhookDeliveryService: Cannot degrade ${subscriptionId}, node not found in Graph.`);
+                return;
             }
+
+            // Degradation is a `status` value, never a `harnessTarget` value. `harnessTarget` is the ROUTING
+            // field, and its documented value space (`mcp-notifications | a2a-webhook | bridge-daemon |
+            // disabled | none`) has no `degraded` member. Both consumers of degradation read
+            // `properties.status`: the sunset check (`ai/scripts/lifecycle/checkSunsetted.mjs`) and the
+            // heartbeat push exclusion (`SwarmHeartbeatService`). Writing 'degraded' into `harnessTarget` left
+            // BOTH blind — the sunset check still classified the dead route as active and heartbeats kept
+            // being pushed at it — while corrupting routing into CoalescingEngineService's unknown-target
+            // drop branch. `upsertNode` merges onto the stored properties, so this touches `status` only.
+            GraphService.upsertNode({
+                id        : subscriptionId,
+                properties: {status: 'degraded'}
+            });
+
+            this.degradedSubscriptions.add(subscriptionId);
+            logger.info(`WebhookDeliveryService: Subscription ${subscriptionId} marked as degraded.`);
         } catch (error) {
             logger.error(`WebhookDeliveryService: Failed to mark subscription ${subscriptionId} degraded: ${error.message}`);
         }
+    }
+
+    /**
+     * @summary The named resumption path out of the degraded state.
+     *
+     * Degradation is terminal by design — a degraded route is never probed again, which is what bounds the
+     * attempt count. Repair is therefore explicit: whoever restores the endpoint clears the marker, and the
+     * persisted `status` must be moved off `'degraded'` in the same operation for the route to survive a
+     * process restart.
+     * @param {String} subscriptionId
+     */
+    clearDegraded(subscriptionId) {
+        this.degradedSubscriptions.delete(subscriptionId);
+        this.consecutiveFailures.delete(subscriptionId);
     }
 }
 

--- a/test/playwright/unit/ai/services/memory-core/WebhookDeliveryService.degradeDeadRoute.spec.mjs
+++ b/test/playwright/unit/ai/services/memory-core/WebhookDeliveryService.degradeDeadRoute.spec.mjs
@@ -1,0 +1,248 @@
+import {setup} from '../../../../setup.mjs';
+
+setup({
+    neoConfig: {
+        unitTestMode: true
+    },
+    appConfig: {
+        name             : 'WebhookDeliveryServiceDegradeDeadRouteTest',
+        isMounted        : () => true,
+        vnodeInitialising: false
+    }
+});
+
+import {test, expect} from '@playwright/test';
+import Neo            from '../../../../../../src/Neo.mjs';
+import * as core      from '../../../../../../src/core/_export.mjs';
+
+/**
+ * A dead wake route that cannot be marked degraded re-runs the full delivery cycle on every subsequent
+ * message forever, with no surfaced signal.
+ *
+ * These specs deliberately do NOT stub `GraphService`. The sibling suite
+ * (`WebhookDeliveryService.spec.mjs`) replaces `getNode` / `upsertNode` with in-memory doubles that
+ * return a node WITH `properties` and never consult row-level security — which healed both halves of
+ * this defect in setup and is why it survived with coverage pointing straight at it. Here the real
+ * service runs against the real `GraphService` read/write path, with the requester left UNBOUND to
+ * reproduce a background flush.
+ *
+ * Every test below fails on the pre-fix tree, and fails for the right reason:
+ * - the degrade never lands, because the RLS-scoped read returns `null` for a node that exists;
+ * - and once it does land, it lands on `harnessTarget`, which no degradation consumer reads.
+ */
+
+// Mutable holder so one fake db can switch the "active requester" between calls. A background timer
+// binds no request, so `null` is the case under test — not an edge case, the normal flush condition.
+const requester = {value: null};
+
+/**
+ * Builds a fake `GraphService.db` holding one pre-warmed WAKE_SUBSCRIPTION node. `autoSave` is left
+ * unset so `upsertNode` mutates the in-memory node without reaching for SQLite, letting the specs
+ * assert against the node object itself.
+ * @param {Object} subscriptionNode
+ * @returns {Object}
+ */
+function makeFakeDb(subscriptionNode) {
+    const nodeMap = new Map([[subscriptionNode.id, subscriptionNode]]);
+
+    return {
+        getAdjacentNodes() {},
+        nodes  : nodeMap,
+        edges  : {getByIndex() { return [] }},
+        storage: {
+            RequestContextService: {
+                getAgentIdentityNodeId: () => requester.value
+            }
+        }
+    };
+}
+
+/**
+ * A WAKE_SUBSCRIPTION exactly as the migration writes it: owner-stamped, not shared, not team-visible —
+ * so every branch of the RLS predicate is false once the requester is absent.
+ * @returns {Object}
+ */
+function makeSubscriptionNode() {
+    return {
+        id        : 'WAKE_SUB:dead-route',
+        label     : 'WAKE_SUBSCRIPTION',
+        properties: {
+            userId               : '@neo-opus-grace',
+            status               : 'active',
+            harnessTarget        : 'a2a-webhook',
+            harnessTargetMetadata: {
+                url       : 'http://127.0.0.1:1/wake',
+                signingKey: 'test-signing-key'
+            }
+        }
+    };
+}
+
+test.describe('WebhookDeliveryService — degrading a dead route from a background flush (#16246)', () => {
+    let GraphService, WebhookDeliveryService, originalDb, originalFetch;
+    let fetchCalls = [];
+
+    test.beforeAll(async () => {
+        GraphService           = (await import('../../../../../../ai/services/memory-core/GraphService.mjs')).default;
+        WebhookDeliveryService = (await import('../../../../../../ai/services/memory-core/WebhookDeliveryService.mjs')).default;
+        originalFetch          = global.fetch;
+    });
+
+    test.beforeEach(() => {
+        originalDb      = GraphService.db;
+        requester.value = null;
+        fetchCalls      = [];
+
+        WebhookDeliveryService.configure({attemptTimeoutSeconds: 1});
+        WebhookDeliveryService.consecutiveFailures.clear();
+        WebhookDeliveryService.degradedSubscriptions?.clear();
+
+        global.fetch = async (url, options) => {
+            fetchCalls.push({url, options});
+            return {ok: false, status: 400};
+        };
+    });
+
+    test.afterEach(() => {
+        GraphService.db = originalDb;
+        requester.value = null;
+        global.fetch    = originalFetch;
+    });
+
+    test('marks a dead route degraded with NO bound request context', async () => {
+        const subscriptionNode = makeSubscriptionNode();
+        GraphService.db        = makeFakeDb(subscriptionNode);
+
+        // Necessity probe (AC 2): the mechanism claims the flush has no resolvable requester. Assert that
+        // directly — if a requester IS bound here, the ticket's mechanism is wrong and the fix is misplaced.
+        expect(GraphService.db.storage.RequestContextService.getAgentIdentityNodeId()).toBe(null);
+
+        // Control: with no requester bound, the RLS-scoped read genuinely cannot see this node. This is the
+        // defect's cause, asserted rather than assumed — and it must stay true, because a fix that works by
+        // weakening RLS would open a cross-tenant read path to close a logging bug.
+        expect(GraphService.getNode({id: subscriptionNode.id})).toBe(null);
+        expect(GraphService.getNodeRecord({id: subscriptionNode.id})).toBe(null);
+
+        const outcome = await WebhookDeliveryService.deliver(subscriptionNode, {eventId: '01HXXX'});
+
+        expect(outcome).toBe('skipped');
+
+        // The assertion that fails on the pre-fix tree: the degrade was skipped entirely, so the route's
+        // state never changed and the next message repeated the whole sequence.
+        expect(subscriptionNode.properties.status).toBe('degraded');
+    });
+
+    test('degradation lands on `status`, the field its consumers actually read', async () => {
+        const subscriptionNode = makeSubscriptionNode();
+        GraphService.db        = makeFakeDb(subscriptionNode);
+
+        await WebhookDeliveryService.deliver(subscriptionNode, {eventId: '01HXXX'});
+
+        // `degraded` is a `status` value; `disabled` is a `harnessTarget` value. Both consumers of
+        // degradation read `properties.status`: the sunset check (`checkSunsetted.mjs` — which otherwise
+        // classifies this dead route as ACTIVE, since `status` is untouched and `harnessTarget` is not
+        // 'disabled') and the heartbeat push exclusion (`SwarmHeartbeatService`, whose SQL predicate is
+        // `$.properties.status != 'degraded'`). Writing into `harnessTarget` left both blind.
+        expect(subscriptionNode.properties.status).toBe('degraded');
+
+        // The routing field keeps its documented value space: mcp-notifications | a2a-webhook |
+        // bridge-daemon | disabled | none. 'degraded' was never a member, and only "worked" by falling
+        // through CoalescingEngineService's unknown-target branch, which drops the digest.
+        expect(subscriptionNode.properties.harnessTarget).toBe('a2a-webhook');
+
+        // The degrade must not disturb anything else — notably the delivery credentials, without which the
+        // route could never be repaired, and the owner stamp, whose loss would make the node null-owned and
+        // therefore readable by every tenant.
+        expect(subscriptionNode.properties.harnessTargetMetadata.url).toBe('http://127.0.0.1:1/wake');
+        expect(subscriptionNode.properties.harnessTargetMetadata.signingKey).toBe('test-signing-key');
+        expect(subscriptionNode.properties.userId).toBe('@neo-opus-grace');
+    });
+
+    test('delivery attempts against a permanently dead route are bounded across repeated messages', async () => {
+        const subscriptionNode = makeSubscriptionNode();
+        GraphService.db        = makeFakeDb(subscriptionNode);
+
+        // Counted over N sends rather than read off the threshold constant: the constant was always 3, and
+        // the defect was that crossing it changed nothing, so every later message paid the full cycle again.
+        for (let i = 0; i < 6; i++) {
+            await WebhookDeliveryService.deliver(subscriptionNode, {eventId: `01HXXX-${i}`});
+        }
+
+        // A 4xx degrades immediately without retrying, so a bounded route costs exactly ONE attempt: the
+        // first. Pre-fix this is 6 — one per message, forever, and 4 apiece once the endpoint refuses the
+        // connection rather than answering 4xx.
+        expect(fetchCalls.length).toBe(1);
+        expect(subscriptionNode.properties.status).toBe('degraded');
+    });
+
+    test('a degraded route is skipped without an attempt, and clearDegraded is the way back', async () => {
+        const subscriptionNode = makeSubscriptionNode();
+
+        subscriptionNode.properties.status = 'degraded';
+        GraphService.db                    = makeFakeDb(subscriptionNode);
+
+        expect(await WebhookDeliveryService.deliver(subscriptionNode, {eventId: '01HXXX'})).toBe('skipped');
+        expect(fetchCalls.length).toBe(0);
+
+        // Terminal by design, with a NAMED resumption path — the property that makes the bound safe to hold.
+        subscriptionNode.properties.status = 'active';
+        WebhookDeliveryService.clearDegraded(subscriptionNode.id);
+
+        global.fetch = async (url, options) => {
+            fetchCalls.push({url, options});
+            return {ok: true, status: 200};
+        };
+
+        expect(await WebhookDeliveryService.deliver(subscriptionNode, {eventId: '01HYYY'})).toBe('delivered');
+        expect(fetchCalls.length).toBe(1);
+    });
+});
+
+test.describe('GraphService.getUnscopedNodeRecord — the one sanctioned context-free read (#16246)', () => {
+    let GraphService, originalDb;
+
+    test.beforeAll(async () => {
+        GraphService = (await import('../../../../../../ai/services/memory-core/GraphService.mjs')).default;
+    });
+
+    test.beforeEach(() => {
+        originalDb      = GraphService.db;
+        requester.value = null;
+    });
+
+    test.afterEach(() => {
+        GraphService.db = originalDb;
+        requester.value = null;
+    });
+
+    test('returns an owner-stamped node that the RLS-scoped reads cannot see', () => {
+        const subscriptionNode = makeSubscriptionNode();
+        GraphService.db        = makeFakeDb(subscriptionNode);
+
+        expect(GraphService.getNodeRecord({id: subscriptionNode.id})).toBe(null);
+
+        const record = GraphService.getUnscopedNodeRecord({
+            id    : subscriptionNode.id,
+            writer: 'spec'
+        });
+
+        expect(record.id).toBe(subscriptionNode.id);
+        expect(record.type).toBe('WAKE_SUBSCRIPTION');
+        expect(record.properties.userId).toBe('@neo-opus-grace');
+    });
+
+    test('returns null for a node that genuinely does not exist', () => {
+        GraphService.db = makeFakeDb(makeSubscriptionNode());
+
+        expect(GraphService.getUnscopedNodeRecord({id: 'WAKE_SUB:absent', writer: 'spec'})).toBe(null);
+    });
+
+    test('refuses a call that does not name its background writer', () => {
+        GraphService.db = makeFakeDb(makeSubscriptionNode());
+
+        // The `writer` name is what keeps this accessor greppable and auditable rather than an anonymous
+        // hole. Making it required means a copy-paste into a new call site cannot stay unattributed.
+        expect(() => GraphService.getUnscopedNodeRecord({id: 'WAKE_SUB:dead-route'}))
+            .toThrow(/requires a `writer` name/);
+    });
+});

--- a/test/playwright/unit/ai/services/memory-core/WebhookDeliveryService.spec.mjs
+++ b/test/playwright/unit/ai/services/memory-core/WebhookDeliveryService.spec.mjs
@@ -27,20 +27,22 @@ let GraphService;
 test.describe('WebhookDeliveryService', () => {
     let fetchCalls   = [];
     let updatedNodes = [];
-    let originalFetch, originalGetNode, originalUpsertNode;
+    let originalFetch, originalGetNode, originalUpsertNode, originalGetUnscopedNodeRecord;
 
     test.beforeAll(async () => {
-        WebhookDeliveryService = (await import('../../../../../../ai/services/memory-core/WebhookDeliveryService.mjs')).default;
-        GraphService           = (await import('../../../../../../ai/services/memory-core/GraphService.mjs')).default;
-        originalFetch          = global.fetch;
-        originalGetNode        = GraphService.getNode;
-        originalUpsertNode     = GraphService.upsertNode;
+        WebhookDeliveryService       = (await import('../../../../../../ai/services/memory-core/WebhookDeliveryService.mjs')).default;
+        GraphService                 = (await import('../../../../../../ai/services/memory-core/GraphService.mjs')).default;
+        originalFetch                = global.fetch;
+        originalGetNode              = GraphService.getNode;
+        originalUpsertNode           = GraphService.upsertNode;
+        originalGetUnscopedNodeRecord = GraphService.getUnscopedNodeRecord;
     });
 
     test.afterAll(() => {
-        global.fetch            = originalFetch;
-        GraphService.getNode    = originalGetNode;
-        GraphService.upsertNode = originalUpsertNode;
+        global.fetch                       = originalFetch;
+        GraphService.getNode               = originalGetNode;
+        GraphService.upsertNode            = originalUpsertNode;
+        GraphService.getUnscopedNodeRecord = originalGetUnscopedNodeRecord;
     });
 
     test.beforeEach(() => {
@@ -75,6 +77,12 @@ test.describe('WebhookDeliveryService', () => {
             return null;
         };
 
+        // The degrade path reads through the context-free accessor, because it runs from a background
+        // flush where the RLS-scoped reads resolve no requester and return null. These doubles
+        // keep this suite's fast in-memory shape; the real read/write path — with RLS actually engaged
+        // and the requester unbound — is covered by WebhookDeliveryService.degradeDeadRoute.spec.mjs.
+        GraphService.getUnscopedNodeRecord = ({id}) => GraphService.getNode({id});
+
         GraphService.upsertNode = (node) => {
             if (node.id === 'WAKE_SUB:123') {
                 updatedNodes.push(node);
@@ -83,6 +91,7 @@ test.describe('WebhookDeliveryService', () => {
 
         // Reset the service state
         WebhookDeliveryService.consecutiveFailures.clear();
+        WebhookDeliveryService.degradedSubscriptions.clear();
         WebhookDeliveryService.configure({attemptTimeoutSeconds: 30});
     });
 
@@ -194,7 +203,9 @@ test.describe('WebhookDeliveryService', () => {
         test.expect(outcome).toBe('skipped');
         test.expect(fetchCalls.length).toBe(1); // No retries for 4xx
         test.expect(updatedNodes.length).toBe(1);
-        test.expect(updatedNodes[0].properties.harnessTarget).toBe('degraded');
+        // `status`, not `harnessTarget`: degradation and routing are separate fields with separate value
+        // spaces, and both consumers of degradation read `status`.
+        test.expect(updatedNodes[0].properties.status).toBe('degraded');
     });
 
     test('retries on 5xx response and degrades after 3 failures', async () => {
@@ -218,7 +229,7 @@ test.describe('WebhookDeliveryService', () => {
         global.setTimeout = originalSetTimeout;
 
         test.expect(fetchCalls.length).toBe(12); // 3 events * 4 attempts
-        test.expect(updatedNodes[0].properties.harnessTarget).toBe('degraded');
+        test.expect(updatedNodes[0].properties.status).toBe('degraded');
     });
 
     test('refuses unsigned Shape-B delivery without issuing a request', async () => {

